core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java

```diff
@@ -109,113 +130,169 @@ public ListBucketResult() {}
+    mIsTruncated = false; // explicitly set the starting value of false
     mMaxKeys = options.getMaxKeys();
+    if (mMaxKeys < 0) {
+      throw new S3Exception(new S3ErrorCode(
+          S3ErrorCode.INVALID_ARGUMENT.getCode(),
+          "MaxKeys may not be negative",
+          S3ErrorCode.INVALID_ARGUMENT.getStatus()));
+    }
     mDelimiter = options.getDelimiter();
-    mEncodingType = options.getEncodingType();
+    if (mDelimiter != null) {
+      mCommonPrefixes = new ArrayList<>();
+    } // otherwise, mCommonPrefixes is null
+    mEncodingType = options.getEncodingType() == null ? ListBucketOptions.DEFAULT_ENCODING_TYPE
+        : options.getEncodingType();
 
     mListType = options.getListType();
-    mContinuationToken = options.getContinuationToken();
-    mStartAfter = options.getStartAfter();
+    if (mListType == null) { // ListObjects v1
+      // Providing a null Marker still results in an empty value being returned,
+      // so we set the default value to an empty string
+      mMarker = options.getMarker() == null ? "" : options.getMarker();
+      mNextMarker = null; // explicitly set the starting value of null
+    } else { // ListObjectsV2
+      mKeyCount = null; // explicitly set the starting value of null
+      mContinuationToken = options.getContinuationToken();
+      mNextContinuationToken = null; // explicitly set the starting value of null
+      mStartAfter = options.getStartAfter();
+    }
 
+    if (mMaxKeys == 0) {
+      return;
+    }
     // contains both ends of "/" character
-    final String mNamePrefix = AlluxioURI.SEPARATOR + mName + AlluxioURI.SEPARATOR;
-    final List<URIStatus> keys = filterKeys(mNamePrefix, children);
-    buildContentsAndCommonPrefix(mNamePrefix, keys);
+    final String mBucketPrefix = AlluxioURI.SEPARATOR + mName + AlluxioURI.SEPARATOR;
+    buildListBucketResult(mBucketPrefix, children);
   }
 
   /**
-   * Filter {@link URIStatus} use marker/continuation-token, prefix and limit.
+   * Filter {@link URIStatus} use marker/continuation-token, prefix, delimiter, and max-keys.
    * @param children a list of {@link URIStatus}, representing the objects and common prefixes
-   * @return A list of {@link URIStatus} after filtering
    */
-  private List<URIStatus> filterKeys(String bucketPrefix, List<URIStatus> children)
-      throws S3Exception {
+  private void buildListBucketResult(
+      String bucketPrefix, List<URIStatus> children) throws S3Exception {
     final String marker;
     if (isVersion2()) {
-      marker = decodeToken(mContinuationToken);
+      if (mContinuationToken != null) {
+        marker = decodeToken(mContinuationToken);
+      } else {
+        marker = "";
+      }
     } else {
       marker = mMarker;
     }
+
+    //group by common prefix if delimiter is provided
+    Set<String> commonPrefixes = new HashSet<>();
+    // used when handling truncating
+    AtomicReference<String> priorNextMarker = new AtomicReference<>();
```

Comment:
I just relied on IntelliJ to suggest how to have a "mutable" String throughout the stream processing. Not sure if this is actually guaranteed to be written to in-order by the stream. @tcrain Any thoughts?

Reply:
This is fine, it doesn't have to be an atomic reference since this is a sequential stream (see the call to Children.stream) so you also use a reference that is not atomic.

Reply:
Resolving as outdated